### PR TITLE
Humberto Antunes

### PR DESCRIPTION
**Pull request by jlouros**

### CODE REVIEW

File: src/Microservice.Api/Controllers/FundsController.cs [hunk 1]
```diff
@@ -34,55 +40,96 @@ public async Task<IEnumerable<FundDetails>> Get()
...
+                    }
 
-                var isValid = Validator.TryValidateObject(fundsDetails, vc, results);
-                _context.Funds.Add(fundMapper.FromContractToDomain(fundsDetails));
-                _context.SaveChangesAsync();
+                    _context.Funds.Add(fund);
+                    _context.SaveChanges();
 
-                return Created("http://localhost:5000/api/funds", fundsDetails.id);
+                    return Created("http://localhost:12564/api/funds", fundsDetails.id);
```
Comment (jlouros) [inline src/Microservice.Api/Controllers/FundsController.cs]: hardcoded location

File: src/Microservice.Api/Controllers/FundsController.cs [hunk 2]
```diff
@@ -34,55 +40,96 @@ public async Task<IEnumerable<FundDetails>> Get()
...
+                            .Include(f => f.AssetBreakdown)
+                                .ThenInclude(f => f.Items)
+                            .Include(f => f.Price)
+                            .Include(f => f.Codes)
+                            .Include(f => f.Yield)
+                            .FirstOrDefault(f => f.Id == id)));
         }
 
+        /// <summary>
+        /// This method receives a fund, validats if they exist, and then validates the fund details.
```
Comment (jlouros) [inline src/Microservice.Api/Controllers/FundsController.cs]: understanding of Swagger documentation 👌

File: src/Microservice.Api/Controllers/ImportController.cs [hunk 3]
```diff
@@ -24,44 +24,100 @@ public ImportController(FundsDbContext context)
...
 
+        /// <summary>
+        /// This method receives a set of fund, validates if they exist and then validates their details
+        /// If a fund exists it is not inserted, and in the end a list of the funds that were already in the database is returned.
+        /// All fund details are validated, if any error is found no fund is saved
+        /// </summary>
+        /// <param name="payload"></param>
+        /// <response code="200">Returns 200 Ok if all fund are successfully saved</response>
+        /// <response code="400">The request body is not in the expected format</response>
+        /// <response code="409">Retuns all the validation error results or the existing funds id</response>
```
Comment (jlouros) [inline src/Microservice.Api/Controllers/ImportController.cs]: good documentation 👌

File: src/Microservice.Api/Controllers/ImportController.cs [hunk 4]
```diff
@@ -24,44 +24,100 @@ public ImportController(FundsDbContext context)
...
-                // TO DO, fail if the fund already exists
+                List<string> existingFundIds = new List<string>();
+
+                FundDetails[] fundsDetails = (payload).ToObject<FundDetails[]>();
+
+                foreach (var fd in fundsDetails)
+                {
+
+                    //Check if a Fund with the provided fund Id already exists in the database
+                    if (_context.Funds.Any(x => x.Id == fd.id))
```
Comment (jlouros) [inline src/Microservice.Api/Controllers/ImportController.cs]: improper string compare

File: src/Microservice.Api/Controllers/ImportController.cs [hunk 5]
```diff
@@ -24,44 +24,100 @@ public ImportController(FundsDbContext context)
...
+        /// </summary>
+        /// <returns></returns>
         [HttpPost]
         [Consumes("application/zip")]
         public async Task<IActionResult> UploadFile()
         {
             using (var sr = new StreamReader(Request.Body))
             {
-
                 using (FileStream fs = new FileStream(Path.Combine(".", "asd.zip"), FileMode.CreateNew, FileAccess.Write))
```
Comment (jlouros) [inline src/Microservice.Api/Controllers/ImportController.cs]: multiple file upload will clash

File: src/Microservice.Api/Helpers/FundValidator.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
+using Microservice.Api.Contracts;
+using Microservice.Api.Domain;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+
+namespace Microservice.Api.Helpers
+{
+    public static class FundValidator
```
Comment (jlouros) [inline src/Microservice.Api/Helpers/FundValidator.cs]: why usage of static method?

**Comment by jlouros** [comment]
### This candidate was unable to solve the following issues:

- did not use dependency injection
- did not use logging
- partial unit test setup (did not fix previous tests)
- did not create any tests
- did not create for his own validation class
- strongly type configuration
- multiple file upload will clash (static file name)

**Comment by oranintelliflo** [comment]
- Imports to same file name
- Returns BadRequest only if payload isn't a JArray
- In validator the naming of the breakdown validation flags is negated. Can be confusing to others who read the code (attention to details)
- Returns 409 if any of the funds are invalid, should be 400 (attention to details)
- Explain the code. What is the purpose of this?
	if (dbProvider != null)
    {
		_context.Entry(dbProvider).State = EntityState.Modified;
    }
- FundsController - Get only returns 200 (optimistic). Possible null exception because using FirstOrDefault and feeding the result straight into the mapper
- Hard coded URL for created resoruce
- Errors in the code (FundMapper.cs)
- BatchFundImportService - no validation of fund details read from files
- CheckArchivesToImport - Code not defensive. Plenty of potential exceptions but no exception handling

**Pull request closed without merging**